gk/fib.c

```diff
@@ -115,12 +180,22 @@ neigh_get_ether_cache_locked(struct neighbor_hash_table *neigh,
 static int
 neigh_del_ether_cache(struct neighbor_hash_table *neigh, void *key)
 {
-	int ret = rte_hash_del_key(neigh->hash_table, key);
-	if (ret >= 0)
-		return initialize_ether_cache(&neigh->cache_tbl[ret]);
+	struct ether_cache *eth_cache;
+	int ret = rte_hash_lookup_data(
+		neigh->hash_table, key, (void **)&eth_cache);
+	if (ret == 0) {
+		int succ = rte_hash_del_key(neigh->hash_table, key);
```

Comment:
If you follow my other comment, this change would become unnecessary. In case you think that we still need this change, keep reading.

We are paying for the cost of two lookups here. First at `rte_hash_lookup_data()` and later at `rte_hash_del_key()`. Let's minimize it calculating the hash only once by using `rte_hash_hash()`, `rte_hash_lookup_with_hash_data()`, and `rte_hash_del_key_with_hash()`.

Reply:
After checking the example, I think in either way (i.e., storing the index of ``eth_cache`` or its address), we still needs to do a lookup to find the data part to delete the entry from the entry table, since the data part actually stores the information about the cached entry.

Reply:
I don't see your point. `rte_hash_del_key()` returns the index of the entry if it exists, or `-ENOEN` if the entry does not exist. So `rte_hash_del_key()` can do, with a single lookup, the work of `rte_hash_lookup_data()`/`rte_hash_lookup_with_hash_data()` and `rte_hash_del_key()`/`rte_hash_del_key_with_hash()`.

Reply:
Here, we use a hash table which stores the ``key-value`` data, and a cache entry table which stores the real Ethernet cached headers. When we call ``rte_hash_add_key_data()`` to add a ``key-value`` pair, (e.g., ``<ip1, index1>``, where ``index1`` represents the index of the cache entry in the cache table), into the hash table with position, say, ``loc1``. Then, later we call ``rte_hash_del_key()`` to delete the entry ``<ip1, index1>``, actually, it returns ``loc1``, that is the location of the ``key-value`` pair, instead of the index of the Ethernet cache entry (i.e., ``index1``). So, we need to first lookup the ``key-value`` pair to find the index of the cache entry, and then delete the ``key-value`` pair from the hash table, and clear the Ethernet cache entry.

Reply:
Reading the source code of DPDK, I've learned that `rte_hash_del_key()` always returns the index. So your assumption that `rte_hash_del_key()` returns `loc1` is not correct. But your approach is correct due to an interesting reason!

Contrary to the documentation (http://dpdk.org/doc/api/rte__hash_8h.html#a31fbd5835f9ed6ff92e052a79f905237),  `rte_hash_add_key_data()` does not return zero when it adds a key, but an index, the same index that `rte_hash_del_key()` will return. Unfortunately, we cannot use this index because we need to set up the entry before we add it to the hash table.

Restating, the current code is fine. Thank you.

Reply:
@cjdoucette, you should read this discussion since you may eventually need to use `rte_hash_add_key_data()`.